Eggsac carrier revival

Birdtalon opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->

[Forum Thread](https://forum.cm-ss13.com/t/eggsac-carrier-revival/4711)

# About the pull request

The concept of this PR is to find a middle ground between the current eggsac carrier and the pre #4459 eggsac carrier.

This PR will make the following changes. (From this point "normal weeds" can be substituted for "off hive weeds" Placing eggs on hive weeds is **unchanged**.)

- Eggsac carrier can once again place eggs on normal weeds.
- Eggsac carrier can only place 4 eggs at once on normal weeds.
- If the Eggsac places more than 4 eggs on normal weeds, their oldest placed egg will die. No hugger release.
- Eggs placed on normal weeds have a maximum lifetime of 5 MINUTES after which they will die. No hugger release.
- Eggs placed on normal weeds have a 1 MINUTE life whilst the carrier is further than 14 tiles away from them.
- If the carrier dies, all of their sustained eggs die.

# Explain why it's good for the game

Eggsac carrier at the moment is in a bit of a poor place and has gone from being very strong to quite poor. Considering the limitation of having to place only on hive weeds.
The majority of feedback I read against eggsac carrier was with the quantity of eggs able to be placed, as well as the locations they can be placed in, all across the map and with impunity.

This PR aims to address those concerns to make the eggsac both less infuriating to play against while still being satisfying to play as a frontline support or as a stealthy trapper.

Eggs can no longer be placed all over the map because of the 4 egg limit off weeds, so the carrier has to think where they want to impact the map. The carrier also has to stay within a reasonable distance to where they are impacting with their eggs which localises their impact to their immediate play area. The carrier also has to be more reactive to current events as they cannot place an egg which then becomes useful 30 minutes later.

Killing the carrier also has a small reward as in addition to removing a xeno from the game, the eggs they are sustaining are cleared. If a carrier is supporting the front and dies, the marines don't then have to clear X number of eggs AFTER the kill.

# Testing Photographs and Procedure

1. Spawn as egg carrier.
2. Plant egg on normal weeds.
3. Move 15+ tiles away.
4. Wait 1 minute
5. OR Wait 5 minutes within 14 tiles.
6. Kill the carrier.

The 5 minute lifetime of the eggs will also clear the eggs in the case that the carrier is admin deleted, or some other weird stuff happens which doesn't result in a death. This will also catch carriers de-evolving

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>


# Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly label your changes in the changelog. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->
<!-- If you add a name after the ':cl', that name will be used in the changelog. You must add your CKEY after the CL if your GitHub name doesn't match. Maintainers freely reserve the right to remove and add tags should they deem it appropriate. -->

:cl:
add: Eggsac carrier can now place eggs on normal weeds to a maximum of 4 eggs.
add: Eggsac carrier eggs on normal weeds have an expiry date.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! -->

Birdtalon commented:
Need to add: handling for eggs bursting for marines or being destroyed as I forgot this. Can't set to draft on mobile github.

Birdtalon commented:
Updated with further handling I missed originally.

fira commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,68 @@
+/obj/effect/alien/egg/carrier_egg
+	name = "fragile egg"
+	desc = "It looks like a weird, fragile egg."
+	var/owner = null
```
Explicit owner type, at last /xenomorph

Birdtalon replied:
Done.

fira commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,68 @@
+	else
+		//Die after maximum lifetime
+		life_timer = addtimer(CALLBACK(src, PROC_REF(remove_owner_and_decay)), CARRIER_EGG_MAXIMUM_LIFE, TIMER_STOPPABLE)
+		set_owner(planter)
```
If you keep a hard reference to owner around you need to be able to clear it up if they die. Eg. using COMSIG_PARENT_QDELETING

Birdtalon replied:
Have added this signal, I also unregister it on `.../Burst()` as I've already setting the owner to null so it should just clean-up any cases where the carrier is deleted. Lmk if any issues with this.

fira commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,68 @@
+	my_delegate.eggs_sustained += src
+	owner = planter
+
+/obj/effect/alien/egg/carrier_egg/proc/remove_owner()
+	if(owner)
+		var/mob/living/carbon/xenomorph/carrier/my_owner = owner
+		var/datum/behavior_delegate/carrier_eggsac/my_delegate = my_owner.behavior_delegate
+		my_delegate.eggs_sustained -= src
+		owner = null
```
i think this would belong better on the behavior_delegate but it's just a detail

Birdtalon replied:
Moved to `behavior_delegate`.

Zonespace27 commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,60 @@
+/obj/effect/alien/egg/carrier_egg
+	name = "fragile egg"
+	desc = "It looks like a weird, fragile egg."
+	var/mob/living/carbon/xenomorph/carrier/owner = null
```
null this on destroy

Zonespace27 commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,60 @@
+		set_owner(planter)
+
+/obj/effect/alien/egg/carrier_egg/Destroy()
+	. = ..()
```
call parent at the end of destroy

Zonespace27 commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,60 @@
+	if(life_timer)
+		deltimer(life_timer)
+
+/obj/effect/alien/egg/carrier_egg/proc/set_owner(mob/living/carbon/xenomorph/carrier/planter)
```
code doc

Zonespace27 commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,60 @@
+	var/datum/behavior_delegate/carrier_eggsac/my_delegate = planter.behavior_delegate
+	my_delegate.eggs_sustained += src
+	owner = planter
+	RegisterSignal(owner, COMSIG_PARENT_QDELETING, PROC_REF(cleanup_owner))
```
you don't need to do this, you can just do the destroy bit i mentioned above

Zonespace27 commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,60 @@
+	UnregisterSignal(owner, COMSIG_PARENT_QDELETING)
+	owner = null
+
+/obj/effect/alien/egg/carrier_egg/proc/cleanup_owner()
```
code doc

Zonespace27 commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/carrier/eggsac.dm:
```diff
@@ -40,8 +43,44 @@
+	var/sustain_distance = 14
+
+/datum/behavior_delegate/carrier_eggsac/append_to_stat()
+	. = "Eggs sustained: [length(eggs_sustained)] / [egg_sustain_cap]"
```
are you sure it's = instead of +=?

Birdtalon replied:
I don't think it actually matters in this case. But comparing it to the same functionality on other behavior delegates I should be consistent with them, also returning it as a list.

https://github.com/Birdtalon/cmss13/blob/c7e8ff476a2fbae3654794173d38f941a2fcc775/code/modules/mob/living/carbon/xenomorph/XenoProcs.dm#L88-L90

Zonespace27 commented:
Willing to give this a shot.

Birdtalon commented:
Think I've addressed all your comments.

- Remove the signal handler
- null owner on destroy
- Added code doc where requested
- Returning list and `+=` on `append_to_stat()`

Zonespace27 commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,55 @@
+/obj/effect/alien/egg/carrier_egg
+	name = "fragile egg"
+	desc = "It looks like a weird, fragile egg."
+	var/mob/living/carbon/xenomorph/carrier/owner = null
+	var/last_refreshed = null
+	var/life_timer = null
```
code doc these

Zonespace27 commented on code/modules/cm_aliens/structures/egg.dm:
```diff
@@ -296,3 +296,55 @@
+	if(life_timer)
+		deltimer(life_timer)
+	owner = null
+	. = ..()
```
```suggestion
	return ..()
```

Zonespace27 commented on code/modules/mob/living/carbon/xenomorph/mutators/strains/carrier/eggsac.dm:
```diff
@@ -40,8 +43,45 @@
 
+/datum/behavior_delegate/carrier_eggsac
+	name = "Eggsac Carrier Behavior Delegate"
+	var/list/eggs_sustained = list()
+	var/egg_sustain_cap = EGGSAC_OFF_WEED_EGGCAP
+	var/sustain_distance = 14
```
code doc

Zonespace27 reviewed (changes requested):
a few nits left

Birdtalon commented:
Addressed your comments.

HeresKozmos commented:
nice!
